### PR TITLE
[Console] Add auto-completion to debug:event-dispatcher

### PR DESCRIPTION
**Pull request by stephenkhoo**
| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | Part of #43594
| License       | MIT
| Doc PR        | -

Adding completion for debug:event-dispatcher.
- [x] event
- [x] --dispatcher 
- [x] --format

Test for 
- [x] event
- [x] --dispatcher 
- [x] --format

### CODE REVIEW

**Comment by GromNaN** [comment]
Hi @stephenkhoo, as [mentioned](https://github.com/symfony/symfony/issues/43594#issuecomment-961299830) by @wouterj, you can can add tests for completion support only.

Do you need help to finish this PR?

**Comment by stephenkhoo** [comment]
Let me try to do one without the test tonight.

File: src/Symfony/Bundle/FrameworkBundle/Command/EventDispatcherDebugCommand.php [hunk 1]
```diff
@@ -120,6 +123,40 @@ protected function execute(InputInterface $input, OutputInterface $output): int
...
+                $dispatcher->getListeners()
+            ));
+
+            return;
+        }
+
+        if ($input->mustSuggestOptionValuesFor('dispatcher')) {
+            $suggestions->suggestValues(array_merge(
+                $this->dispatchers->getServiceIds(),
+                $this->dispatchers->getAliases()
```
Comment (stephenkhoo) [inline src/Symfony/Bundle/FrameworkBundle/Command/EventDispatcherDebugCommand.php:152]: Not quite sure the dispatchers will be an instance of Container or ContainerBuilder?
Will there be aliases need to be supported on the completion?

**Comment by rvanlaak** [comment]
https://github.com/symfony/symfony/pull/43850/files#diff-5508b4b757a2fd5f0cb21419dd4dbd1478196d16183e023f10696e81af02e9e9R116-R140 could help you in adding a test.

**Comment by stephenkhoo** [comment]
> https://github.com/symfony/symfony/pull/43850/files#diff-5508b4b757a2fd5f0cb21419dd4dbd1478196d16183e023f10696e81af02e9e9R116-R140 could help you in adding a test.

I think for completion test part I've no problem, but I'm not sure what is the best way to inject the $dispatchers which is a class that has interface of Container.
I'll also need the container to have a few dispatchers so that it will be able to provide some suggestions.

**Comment by stephenkhoo** [comment]
> Hi @stephenkhoo, as [mentioned](https://github.com/symfony/symfony/issues/43594#issuecomment-961299830) by @wouterj, you can can add tests for completion support only.
> 
> Do you need help to finish this PR?

I think I'll need some help on direction to write test for the completion.
So far based on what I did on translation update I think they're mostly the same, just that when I need to DI for the dispatchers, I'm quite clueless.

I think if I have clue on how to pass the required parameter for `EventDispatcherDebugCommand` then I should be able to proceed on the completion test.

**Comment by stephenkhoo** [comment]
I've added the tests, there're still a few doubt.
Currently in order to get the dispatcher's name, will use getServiceIds, and getAliases which will not exists in Container interface.

Do we need to update the Type in the Command class?

File: src/Symfony/Bundle/FrameworkBundle/Tests/Command/EventDispatcherDebugCommandCompletionTest.php [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+use Symfony\Component\HttpKernel\KernelInterface;
+
+class EventDispatcherDebugCommandCompletionTest extends TestCase
+{
+    /**
+     * @dataProvider provideCompletionSuggestions
+     */
+    public function testComplete(array $input, array $expectedSuggestions)
+    {
+        $tester = $this->createCommandCompletionTester();
```
Comment (rvanlaak) [inline src/Symfony/Bundle/FrameworkBundle/Tests/Command/EventDispatcherDebugCommandCompletionTest.php:30]: Did you think about only mocking the `ContainerInterface` here (so no kernel, no application)? 

```php
$dispatchers = $this->createMock(ContainerInterface::class);
$dispatchers->method('get')->willReturn($dispatcher);
$dispatchers->method('has')->willReturn(true);
// ...

$command = new EventDispatcherDebugCommand($dispatchers);
$tester = new CommandCompletionTester($command);
```

File: src/Symfony/Bundle/FrameworkBundle/Command/EventDispatcherDebugCommand.php (hunk 1 above)
Comment (rvanlaak) [reply src/Symfony/Bundle/FrameworkBundle/Command/EventDispatcherDebugCommand.php:152]: The Psalm CI job might fail on `getServiceIds` and `getAliases`. Maybe you can attach XDebug to your CLI job?

File: src/Symfony/Bundle/FrameworkBundle/Tests/Command/EventDispatcherDebugCommandCompletionTest.php (hunk 2 above)
Comment (stephenkhoo) [reply src/Symfony/Bundle/FrameworkBundle/Tests/Command/EventDispatcherDebugCommandCompletionTest.php:30]: Let me try this out

Comment (stephenkhoo) [reply src/Symfony/Bundle/FrameworkBundle/Tests/Command/EventDispatcherDebugCommandCompletionTest.php:30]: But container interface don't have another method required
`getServiceIds`, `getAliases`
Will mocking this interface able to make method for methods not on the interface?

File: src/Symfony/Bundle/FrameworkBundle/Command/EventDispatcherDebugCommand.php [hunk 3]
```diff
@@ -120,6 +123,47 @@ protected function execute(InputInterface $input, OutputInterface $output): int
...
+            $listeners = $dispatcher->getListeners();
+            $eventNames = array_keys($listeners);
+            $listenersName = [];
+            foreach ($listeners as $listOfListener) {
+                $listenersName = array_merge($listenersName, $listOfListener);
+            }
+            $suggestions->suggestValues(array_merge(
+                $eventNames,
+                $listenersName,
+            ));
```
Comment (GromNaN) [inline src/Symfony/Bundle/FrameworkBundle/Command/EventDispatcherDebugCommand.php:144]: Why listeners are added to the list of suggested event names? The argument get values like `kernel.response` or `Symfony\Component\Mailer\Event\MessageEvent`. I did some tests and listener names are not accepted (ex: `Symfony\Component\HttpKernel\EventListener\ResponseListener::onKernelResponse`)

```suggestion
            $suggestions->suggestValues(array_keys($dispatcher->getListeners()));
```

File: src/Symfony/Bundle/FrameworkBundle/Command/EventDispatcherDebugCommand.php (hunk 1 above)
Comment (GromNaN) [reply src/Symfony/Bundle/FrameworkBundle/Command/EventDispatcherDebugCommand.php:152]: `$dispatchers` is an instance of `Symfony\Contracts\Service\ServiceProviderInterface`. The method `getProvidedServices()` can be used after checking its availability.

**Review by GromNaN — CHANGES_REQUESTED** [review]
Thank you for your time on this PR and all your attempts to add tests. In order to finish this one before the release of 5.4, I made a PR on your repository. After reviewing, you can merge it and squash the commits. https://github.com/GromNaN/symfony/commit/5b66228a19841e530ee63348ef2735441ab6aa82

**Pull request closed without merging**